File: npc/job_matching.go
// Package npc：求職撮合（10.15），無職且鎂低於閾值的 NPC 與有職缺的場所匹配，寫入 assignment 與排班。對齊討論 002。
package npc

import (
	"math/rand"

	"singularity_world/db"
)

// SeekJobMgThresholdDefault 鎂閾值預設值（config 為 0 時使用）。
const SeekJobMgThresholdDefault = 50

// JobMatchStableProbability 安定需求：無職且鎂≥閾值者參與撮合的機率（細目 3）。
const JobMatchStableProbability = 0.3

// DefaultOccupationForHire 撮合時寫入的職業；場所若有既有指派則取該場所第一種職業（GetFirstOccupationIDForVenue），無則用此常數。
const DefaultOccupationForHire = "服務生"

// 撮合後排班預設班次（日班 06–19）。
const defaultShiftStart, defaultShiftEnd = 6, 19

// JobMatchParams 求職撮合參數（由 main 從 config 填入）。
type JobMatchParams struct {
	MaxPerVenue        int  // 全局限額；單一場所可用 GetVenueMaxStaff 覆寫
	MgThreshold        int  // 鎂低於此才參與；0 用 SeekJobMgThresholdDefault
	JobMatchWhenStable bool // 為 true 時鎂≥閾值者以機率參與
}

// RunJobMatchingTick 執行一輪求職撮合。細目 2：每場所用 GetVenueMaxStaff 取上限；細目 3：threshold、JobMatchWhenStable；細目 5：g 非 nil 時依 NPC 與場所距離優先匹配。回傳本輪新入職的 entity ID 列表。
func RunJobMatchingTick(g *db.RoomGraph, p JobMatchParams) []string {
	if p.MaxPerVenue <= 0 {
		p.MaxPerVenue = 2
	}
	threshold := p.MgThreshold
	if threshold <= 0 {
		threshold = SeekJobMgThresholdDefault
	}
	var added []string
	ids, err := db.GetNPCIDsWithRoom()
	if err != nil || len(ids) == 0 {
		return added
	}
	var unassigned []string
	for _, entityID := range ids {
		list, _ := db.GetAssignmentsForEntity(entityID)
		if len(list) > 0 {
			continue
		}
		c, _ := db.GetEntity(entityID)
		if c == nil {
			continue
		}
		if c.Magnesium < threshold {
			unassigned = append(unassigned, entityID)
		} else if p.JobMatchWhenStable && rand.Float32() < JobMatchStableProbability {
			unassigned = append(unassigned, entityID)
		}
	}
	if len(unassigned) == 0 {
		return added
	}
	venueIDs, err := db.GetAllVenueIDs()
	if err != nil || len(venueIDs) == 0 {
		return added
	}
	var slots []string
	for _, vid := range venueIDs {
		n, _ := db.GetAssignmentCountByVenue(vid)
		limit := db.GetVenueMaxStaff(vid, p.MaxPerVenue)
		for k := 0; k < limit-n; k++ {
			slots = append(slots, vid)
		}
	}
	if len(slots) == 0 {
		return added
	}
	spawnRoom := db.GetSpawnRoomID()
	if g != nil {
		rand.Shuffle(len(slots), func(i, j int) { slots[i], slots[j] = slots[j], slots[i] })
		for _, venueID := range slots {
			roomIDs, _ := db.GetRoomIDsForVenue(venueID)
			if len(roomIDs) == 0 {
				continue
			}
			venueFirst := roomIDs[0]
			bestIdx := -1
			bestDist := 9999
			for i, entityID := range unassigned {
				current, _ := db.GetEntityRoom(entityID)
				if current == "" {
					continue
				}
				path := g.FindPath(current, venueFirst)
				dist := len(path)
				if current == venueFirst {
					dist = 0
				}
				if dist < bestDist {
					bestDist = dist
					bestIdx = i
				}
			}
			if bestIdx < 0 {
				continue
			}
			entityID := unassigned[bestIdx]
			unassigned = append(unassigned[:bestIdx], unassigned[bestIdx+1:]...)
			if doAssign(entityID, venueID, spawnRoom) {
				added = append(added, entityID)
			}
			if len(unassigned) == 0 {
				break
			}
		}
	} else {
		rand.Shuffle(len(unassigned), func(i, j int) { unassigned[i], unassigned[j] = unassigned[j], unassigned[i] })
		rand.Shuffle(len(slots), func(i, j int) { slots[i], slots[j] = slots[j], slots[i] })
		n := len(unassigned)
		if len(slots) < n {
			n = len(slots)
		}
		for i := 0; i < n; i++ {
			if doAssign(unassigned[i], slots[i], spawnRoom) {
				added = append(added, unassigned[i])
			}
		}
	}
	return added
}

func doAssign(entityID, venueID, spawnRoom string) bool {
	occupation := db.GetFirstOccupationIDForVenue(venueID)
	if occupation == "" {
		occupation = DefaultOccupationForHire
	}
	if db.InsertAssignment(entityID, occupation, venueID, "job_matching") != nil {
		return false
	}
	roomIDs, _ := db.GetRoomIDsForVenue(venueID)
	var workRoom string
	if len(roomIDs) > 0 {
		workRoom = roomIDs[0]
	}
	if workRoom != "" && spawnRoom != "" {
		_ = db.InsertSchedule(entityID, workRoom, spawnRoom, defaultShiftStart, defaultShiftEnd)
	}
	return true
}

File: src/npc/job_matching.rs
//! 求職撮合（對齊既有 `npc/job_matching`）。

use rand::Rng;

use crate::db::{
    get_all_venue_ids, get_assignment_count_by_venue, get_assignments_for_entity, get_entity,
    get_entity_room, get_first_occupation_id_for_venue, get_npc_ids_with_room, get_room_ids_for_venue,
    get_spawn_room_id, get_venue_max_staff, insert_assignment, insert_schedule, RoomGraph,
};

/// 鎂閾值預設（config 為 0 時使用）。
pub const SEEK_JOB_MG_THRESHOLD_DEFAULT: i32 = 50;
/// 鎂≥閾值且開啟穩定撮合時，參與機率。
pub const JOB_MATCH_STABLE_PROBABILITY: f32 = 0.3;
/// 無既有職業時寫入的預設職業 id。
pub const DEFAULT_OCCUPATION_FOR_HIRE: &str = "服務生";
const DEFAULT_SHIFT_START: i32 = 6;
const DEFAULT_SHIFT_END: i32 = 19;

/// 撮合參數（對齊既有 `JobMatchParams`）。
#[derive(Debug, Clone)]
pub struct JobMatchParams {
    pub max_per_venue: i32,
    pub mg_threshold: i32,
    pub job_match_when_stable: bool,
}

/// Fisher–Yates 洗牌。
fn shuffle_vec<T>(v: &mut [T], rng: &mut impl Rng) {
    for i in (1..v.len()).rev() {
        let j = rng.random_range(0..=i);
        v.swap(i, j);
    }
}

fn do_assign(entity_id: &str, venue_id: &str, spawn_room: &str) -> bool {
    let mut occupation = get_first_occupation_id_for_venue(venue_id);
    if occupation.is_empty() {
        occupation = DEFAULT_OCCUPATION_FOR_HIRE.to_string();
    }
    if insert_assignment(entity_id, &occupation, venue_id, "job_matching").is_err() {
        return false;
    }
    let room_ids = get_room_ids_for_venue(venue_id)
        .ok()
        .flatten()
        .unwrap_or_default();
    let work_room = room_ids.first().cloned().unwrap_or_default();
    if !work_room.is_empty() && !spawn_room.is_empty() {
        let _ = insert_schedule(
            entity_id,
            &work_room,
            spawn_room,
            DEFAULT_SHIFT_START,
            DEFAULT_SHIFT_END,
        );
    }
    true
}

/// 一輪求職撮合；`graph` 為 `None` 時改隨機配對（與既有行為一致）。回傳本輪新入職 entity id。
#[must_use]
pub fn run_job_matching_tick(graph: Option<&RoomGraph>, mut p: JobMatchParams) -> Vec<String> {
    let mut added = Vec::new();
    if p.max_per_venue <= 0 {
        p.max_per_venue = 2;
    }
    let mut threshold = p.mg_threshold;
    if threshold <= 0 {
        threshold = SEEK_JOB_MG_THRESHOLD_DEFAULT;
    }

    let ids = get_npc_ids_with_room();
    if ids.is_empty() {
        return added;
    }

    let mut rng = rand::rng();
    let mut unassigned: Vec<String> = Vec::new();
    for entity_id in ids {
        let Ok(list) = get_assignments_for_entity(&entity_id) else {
            continue;
        };
        if !list.is_empty() {
            continue;
        }
        let Ok(Some(ch)) = get_entity(&entity_id) else {
            continue;
        };
        if ch.magnesium < threshold
            || (p.job_match_when_stable && rng.random_bool(f64::from(JOB_MATCH_STABLE_PROBABILITY)))
        {
            unassigned.push(entity_id);
        }
    }
    if unassigned.is_empty() {
        return added;
    }

    let Ok(venue_ids) = get_all_venue_ids() else {
        return added;
    };
    if venue_ids.is_empty() {
        return added;
    }

    let mut slots: Vec<String> = Vec::new();
    for vid in venue_ids {
        let Ok(n) = get_assignment_count_by_venue(&vid) else {
            continue;
        };
        let limit = get_venue_max_staff(&vid, p.max_per_venue);
        for _ in 0..limit.saturating_sub(n as i32) {
            slots.push(vid.clone());
        }
    }
    if slots.is_empty() {
        return added;
    }

    let spawn_room = get_spawn_room_id();

    if let Some(g) = graph {
        shuffle_vec(&mut slots, &mut rng);
        let mut un = unassigned;
        for venue_id in slots {
            let room_ids = get_room_ids_for_venue(&venue_id)
                .ok()
                .flatten()
                .unwrap_or_default();
            if room_ids.is_empty() {
                continue;
            }
            let venue_first = room_ids[0].clone();
            let mut best_idx: Option<usize> = None;
            let mut best_dist = 9999i32;
            for (i, entity_id) in un.iter().enumerate() {
                let Ok(current) = get_entity_room(entity_id) else {
                    continue;
                };
                if current.is_empty() {
                    continue;
                }
                let dist = if current == venue_first {
                    0
                } else if let Some(path) = g.find_path(&current, &venue_first) {
                    path.len() as i32
                } else {
                    9999
                };
                if dist < best_dist {
                    best_dist = dist;
                    best_idx = Some(i);
                }
            }
            let Some(idx) = best_idx else {
                continue;
            };
            let entity_id = un.remove(idx);
            if do_assign(&entity_id, &venue_id, &spawn_room) {
                added.push(entity_id);
            }
            if un.is_empty() {
                break;
            }
        }
    } else {
        shuffle_vec(&mut unassigned, &mut rng);
        shuffle_vec(&mut slots, &mut rng);
        let n = unassigned.len().min(slots.len());
        for i in 0..n {
            if do_assign(&unassigned[i], &slots[i], &spawn_room) {
                added.push(unassigned[i].clone());
            }
        }
    }

    added
}
